p4-16/psa/PSA.mdk

```diff
@@ -2576,3 +2576,135 @@ Certainly if there are long durations of time when egress decides to
 drop all packets to an output port, that port will go idle.  The
 scheduling algorithm implementations are all built with a finite
 maximum packet scheduling rate.
+
+
+# Appendix: Multi-pipeline PSA devices
+
+The highest packet rate network devices today are ASICs running at a
+clock rate on the order of 1 to 2 GHz.  This discussion will assume 1
+GHz for the sake of a concrete numerical example, but everything
+discussed here scales linearly with the clock rate.
+
+It is common to design a portion of a network ASIC such that it can
+start processing a new packet once every clock cycle, and finish a
+packet every clock cycle.  The latency might be hundreds of clock
+cycles from starting a packet until it is complete.  P4 tables in such
+ASICs are typically implemented using logic such as TCAMs and SRAMs.
+TCAM designs can do 1 search per clock cycle.  The lowest area and
+power SRAMs can do 1 read or 1 write per clock cycle.
+
+While there are "multi ported" SRAM designs that can be read and/or
+written multiple times per clock cycle, these have a noticeable
+increase in area and power over the "single ported" designs that are
+limited to 1 access per cycle.  If multi ported TCAM designs even
+exist, the cost premium for multi-ported TCAMs is likely to be even
+higher than the cost premium for multi-ported SRAMs.  The typical way
+to achieve higher TCAM search rates is via parallelism, by creating
+multiple copies of the desired TCAM, which is a linear increase in
+area and power (at least).
+
+Due to these issues, if one wishes to create a switch ASIC that
+achieves a packet processing rate of N times the clock rate, e.g. 2
+billion packets per second in a 1 GHz ASIC, the most straightforward
+way to do this is to take advantage of the fact that packet processing
+is (mostly) an embarrasingly parallel
+problem[^EmbarrassinglyParallel], and create a device with N
+pipelines[^PipelineNote], each pipeline processing packets at 1
+billion packets per second.
+
+[^EmbarrassinglyParallel]: https://en.wikipedia.org/wiki/Embarrassingly_parallel
+
+[^PipelineNote]: Here and elsewhere in P4 specification documents, the term "pipeline" is often used for such a thing.  This is by now traditional in P4 specifications, and although we will not try to change that here, note that there are other reasonable hardware designs that can accomplish this goal, e.g. a collection of CPU cores running in parallel, each processing different packets.
```

Comment:
"used for such a thing"? Maybe: used to refer to the sequence of blocks operating on the packet?

Reply:
Well "sequence" has similar connotations as a pipeline.  P4 implementations need not be implemented as a pipeline.  I am pretty sure that this name is used because of the RMT architecture being a hardware pipeline design.